keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1344 @@
+- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
+- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))
+
+Additionally, some pod failures are not linked with the container execution,
+but rather with the internal kubernetes cluster management (see:
+[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).
+Such pod failures should be recongnized as infrastructure failures and it
+should be possible to ignore them from the counter towards `backoffLimit`.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Extension of Job API with user-friendly syntax to terminate jobs based on the
+  end state of the failed pod.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Implementation of other indicators of non-retriable jobs such as termination logs.
+- Modification of the semantics for job termination. In particular, allowing for
+  all indexes of an indexed-job to execute when only one or a few indexes fail
+  [#109712](https://github.com/kubernetes/kubernetes/issues/109712).
+- Similar termination policies for other workload controllers such as Deployments
+  or StatefulSets.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Extension of the Job API with a new field which allows to configure the set of
+conditions and associated actions which determine how a pod failure is handled.
+The extended Job API supports discrimination of pod failures based on the
+container exit codes as well as based on the end state of a failed pod.
+
+In order to support discrimination of pod failures based on their end state
+we use the already existing `status.conditions` field to append a dedicated Pod
+Condition indicating (with its type) that the pod is being terminated by an
+internal kubernetes component. Moreover, we modify the internal kubernetes
+components to send an API call to append the dedicated Pod condition along with
+sending the associated Pod delete request. In particluar, the following
+kubernetes components will be modified:
+- kube-controller-manager (taint manager performing pod eviction, also: podgc, node-lifecycle-controller)
+- kube-scheduler (when performing `Preemption`)
+- kubelet (when performing `NodePressure` eviction)
```

Comment:
For Kubelet, you may also want to consider the node-level admission failures that lead to a pod ending up in status.phase == Failed (e.g. status.reason set to `OutOfCpu`, `OutOfMemory`, `UnexpectedAdmissionErrorPod`, etc)

Reply:
We remove kubelet modifications from the KEP (at least for Alpha). The reason being that in many scenarios it is hard to be helpful in recognizing if a pod failure initiated by kubelet should be retried or not. For example, when kubelet initiates node pressure eviction due to memory it may mean either a software bug (in which case we want to terminate the job) or the node being low on memory due to other processes.

Reply:
Admission failures are tricky in a different way. They happen as a result of kube-scheduler and kubelet being out of sync, which can happen when using static pods or when node labels change. So they are kind of a termination by the control plane.
But overall it sounds like there is a lot of debate on what the failures from kubelet are, so we could benefit from an alpha release without accounting for them so we can gather feedback from users.